### PR TITLE
Add support for IRM (Insert Replace Mode)

### LINKED ISSUE #1947: Add support for VT102 Insertion-Replace Mode (IRM)
### Insertion-Replacement Mode (IRM)

```
ESC  [   4   h
033 133 064 150
```

Set selects insert mode and turns INSERT on. New display characters move old display characters to the right. Characters moved past the right margin are lost.

```
ESC  [   4   l
033 133 064 154
```

Reset selects replace mode and turns INSERT off. New display characters replace old display characters at cursor position. The old character is erased.

---

#### original issue body follows.

The following is from recording command line editing in an ssh session to a freebsd environment (via FreeBSD's script). The displayed text is odd in Windows Terminal. First I show a textual form that shows the escape sequences and such in a readable form (I eliminated bell-ring characters from the file and other lines of text):

> more ~/odd_commnd_line_editing_display.txt
> ls ~/sys_typescriptsESC[1ESC[4h-ESC[4lESC[4hlESC[4lESC[4hTESC[4lESC[4hdESC[4lESC[4htESC[4lESC[4h ESC[4

A normal display in other environments (for example macOS terminal doing an ssh into FreeBSD) looks like:

> cat ~/odd_commnd_line_editing_display.txt
> ls -lTdt ~/sys_typescripts/

(The above shows the actual ls command that was executed.)

But under Windows Terminal (Preview) it shows:

> cat ~/odd_commnd_line_editing_display.txt
> ls -lTdt ~ypescripts/

This is also how it ends up looking during the command line editing,
not showing the actual text of the command to be executed: missing
some text.

The sequence is from deciding to add "-lTdt" after having typed the
path. (It was a "set -o vi" type of command line editing context.)

I was using TERM being xterm-256color on FreeBSD .

### PR DESCRIPTION
**Pull request by j4james**
This PR add support for the ANSI Insert/Replace mode (`IRM`), which
determines whether output characters are inserted at the active cursor
position, moving existing content to the right, or whether they should
overwrite the content that is already there.

The implementation is a bit of a hack. When that mode is enabled, it
first measures how many cells the string is expected to occupy, then
scrolls the target line right by that amount before writing out the new
text.

In the longer term it might be better if this was implemented entirely
in the `TextBuffer` itself, so the scrolling could take place at the
same time as the content was being written.

## Validation Steps Performed

I've added a very basic unit test that verifies the mode is working as
expected. But I've also done a lot more manual testing, confirming edge
cases like wide characters, double-width lines, and both with and
without wrapping mode enabled.

Closes #1947

### CODE REVIEW

**Comment by lhecker** [comment]
Here's my approach: https://github.com/microsoft/terminal/compare/dev/lhecker/1947-irm
It contains a lot more fluff than your PR, which is why I prefer yours. `TextBuffer::InsertLine` is the relevant function.

The reason it contains more "fluff" is because I initially tried to implement IRM with proper Unicode support in mind (which means that we need to avoid remeasuring strings) and resulted in a rather complex changeset since it requires a generic text replacement _and_ insertion function. To get it done faster, I removed that code and replaced it with a simpler one that remeasures strings via `OutputCellIterator`. Still, two new functions remained: `ROW::GetText` and `ROW:CopyTo`. But I need the former in the future anyways in order to be able to remove our many buffer iterators (and simplifying text extraction) and the latter to fix #14696 & to implement buffer snapshotting.

...but I failed to implement moving the text attributes until now, whereas your code already (presumably) works flawlessly, while being a rather simple changeset as well. I like it!

**Review by lhecker — APPROVED** [review]

**Comment by j4james** [comment]
@lhecker Your branch is doing almost exactly what I hoped you'd be doing. So if we merge this PR first, it would ideal if you could do a follow up at some point that just replaces my `_ScrollRectHorizontally` hack with your `InsertLine` call once it's ready, because I'm assuming it should be a lot more efficient.

I haven't looked at your actual implementation in much detail, but the one key thing that I think is still missing from the `InsertLine` interface is the ability to specify the line width. It's important that `AdaptDispatch` has control of that in order to support margins in the future.

And I _think_ we might also need to be able to pass in the wrap flag. It's not obvious what the expected behavior should be with IRM enabled, but it seemed reasonable to me that we should handle that the same as replace mode. I should probably test on some other terminals and see what they do.

**Review by zadjii-msft — APPROVED** [review]

**Comment by DHowett** [comment]
> but it seemed reasonable to me that we should handle that the same as replace mode. I should probably test on some other terminals and see what they do.

Agreed. I originally looked at how gnome-terminal handled it, and i wasn't terribly surprised by the results. If I recall, it kept the wrap flag on rows up until you got to the end of the line, at which point normal newline/wrapping behavior took over.

File: src/terminal/adapter/DispatchTypes.hpp [hunk 1]
```diff
@@ -395,6 +395,7 @@ namespace Microsoft::Console::VirtualTerminal::DispatchTypes
 
     enum ModeParams : VTInt
     {
+        IRM_InsertReplaceMode = ANSIStandardMode(4),
```
Comment (DHowett) [inline src/terminal/adapter/DispatchTypes.hpp:398]: I love that my weird "tagged enum" thing didn't turn into YAGNI. I added it _because_ we'd eventually do IRM, and need `SM`/`RM`. 😄

File: src/terminal/parser/telemetry.cpp [hunk 2]
```diff
@@ -221,7 +221,9 @@ void TermTelemetry::WriteFinalTraceLog() const
                                       TraceLoggingUInt32(_uiTimesUsed[SGR], "SGR"),
                                       TraceLoggingUInt32(_uiTimesUsed[DECSC], "DECSC"),
                                       TraceLoggingUInt32(_uiTimesUsed[DECRC], "DECRC"),
+                                      TraceLoggingUInt32(_uiTimesUsed[DECSET], "SM"),
```
Comment (DHowett) [inline src/terminal/parser/telemetry.cpp:224]:
```suggestion
                                      TraceLoggingUInt32(_uiTimesUsed[SM], "SM"),
```

File: src/terminal/parser/telemetry.cpp [hunk 3]
```diff
@@ -221,7 +221,9 @@ void TermTelemetry::WriteFinalTraceLog() const
                                       TraceLoggingUInt32(_uiTimesUsed[SGR], "SGR"),
                                       TraceLoggingUInt32(_uiTimesUsed[DECSC], "DECSC"),
                                       TraceLoggingUInt32(_uiTimesUsed[DECRC], "DECRC"),
+                                      TraceLoggingUInt32(_uiTimesUsed[DECSET], "SM"),
                                       TraceLoggingUInt32(_uiTimesUsed[DECSET], "DECSET"),
+                                      TraceLoggingUInt32(_uiTimesUsed[DECRST], "RM"),
```
Comment (DHowett) [inline src/terminal/parser/telemetry.cpp:226]:
```suggestion
                                      TraceLoggingUInt32(_uiTimesUsed[RM], "RM"),
```

(if I understand this correctly)

**Review by DHowett — APPROVED** [review]
Minor nit, but I'm just blown away by how the foundation you've been laying over the past year has made this look so easy. Leonard as well, of course 🙂

File: src/terminal/parser/telemetry.cpp (hunk 3 above)
Comment (j4james) [reply src/terminal/parser/telemetry.cpp]: I don't know how I missed that. Thanks!
